File: src/main/java/com/mario6/paper/repository/JsonFileProjectRepositoryImpl.java
package com.mario6.paper.repository;

import com.alibaba.fastjson.JSONObject;
import com.alibaba.fastjson.JSONReader;
import com.alibaba.fastjson.serializer.SerializerFeature;
import com.mario6.paper.core.PaperProperties;
import com.mario6.paper.model.Project;
import com.mario6.paper.model.ProjectDataWrapper;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Component;

import javax.annotation.Resource;
import java.io.*;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * JsonFileProjectRepositoryImpl
 *
 * @author Mario Luo
 * @date 2019.01.19 11:53
 */
@Component
public class JsonFileProjectRepositoryImpl implements ProjectRepository {

    @Resource
    private PaperProperties configuration;

    @Override
    public Project findOne(String id) {
        ProjectDataWrapper data = readData();
        return doGetProjectById(data, id);
    }

    @Override
    public int save(Project project) {
        ProjectDataWrapper data = readData();
        List<Project> projects = data.getProjects();
        Project orgProject = doGetProjectById(data, project.getId());
        if(orgProject != null) {
            orgProject.setUpdateTime(project.getUpdateTime());
            orgProject.setCreator(project.getCreator());
            orgProject.setDescription(project.getDescription());
            orgProject.setName(project.getName());
            orgProject.setUploaded(project.getUploaded());
            orgProject.setEntranceUri(project.getEntranceUri());
        } else {
            projects.add(0, project);
        }
        writeData(data);
        return 1;
    }

    @Override
    public List<Project> findAll() {
        ProjectDataWrapper data = readData();
        return data.getProjects();
    }

    @Override
    public int delete(String id) {
        ProjectDataWrapper data = readData();
        List<Project> projects = data.getProjects();

        int count = 0;
        Iterator<Project> it = projects.iterator();
        while(it.hasNext()) {
            Project one = it.next();
            String pid = one.getId();
            if(StringUtils.equals(id, pid)) {
                it.remove();
                count ++;
            }
        }
        return count;
    }

    private Project doGetProjectById(ProjectDataWrapper data, String id) {
        List<Project> projects = data.getProjects();
        for(Project one: projects) {
            if(one == null) continue;
            if(StringUtils.equals(id, one.getId())) {
                return one;
            }
        }
        return null;
    }

    private void writeData(ProjectDataWrapper data) {
        File file = getProjectMetaJsonFile();
        doWriteJsonFile(file, data);
    }

    private ProjectDataWrapper readData() {
        File file = getProjectMetaJsonFile();
        try (Reader reader = new FileReader(file)) {
            JSONReader jsonReader = new JSONReader(reader);
            ProjectDataWrapper data = jsonReader.readObject(ProjectDataWrapper.class);
            if(data.getProjects() == null) {
                data.setProjects(new ArrayList<Project>());
            }
            return data;
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private File getProjectMetaJsonFile() {
        String workDir = configuration.getWorkDir();
        String path = workDir + File.separator + PaperProperties.PROJECT_META_FILE_NAME;
        File file = new File(path);
        if(!file.exists()) {
            File parentFile = file.getParentFile();
            if(!parentFile.exists()) {
                parentFile.mkdirs();
            }
            ProjectDataWrapper data = new ProjectDataWrapper();
            data.setProjects(new ArrayList<Project>());
            doWriteJsonFile(file, data);
        }
        return file;
    }

    private void doWriteJsonFile(File file, ProjectDataWrapper data) {
        try(FileWriter writer = new FileWriter(file)) {
            JSONObject.writeJSONString(writer, data,
                SerializerFeature.WriteNullListAsEmpty,
                SerializerFeature.PrettyFormat,
                SerializerFeature.SortField);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/com/github/lkqm/paper/repository/JsonFileProjectRepositoryImpl.kt
package com.github.lkqm.paper.repository

import com.alibaba.fastjson.JSONObject
import com.alibaba.fastjson.JSONReader
import com.alibaba.fastjson.serializer.SerializerFeature
import com.github.lkqm.paper.config.PaperProperties
import com.github.lkqm.paper.model.Project
import com.github.lkqm.paper.model.ProjectDataWrapper
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component
import java.io.*
import java.util.*
import javax.annotation.Resource

/**
 * JsonFileProjectRepositoryImpl
 *
 * @author Mario Luo
 * @date 2019.01.19 11:53
 */
@Component
class JsonFileProjectRepositoryImpl : ProjectRepository {

    @Resource
    private lateinit var configuration: PaperProperties

    override fun findOne(id: String): Project? {
        val data = readData()
        return doGetProjectById(data, id)
    }

    override fun save(project: Project): Int {
        val data = readData()
        val projects = data.projects
        val orgProject = doGetProjectById(data, project.id)
        if (orgProject != null) {
            orgProject.updateTime = project.updateTime
            orgProject.creator = project.creator
            orgProject.description = project.description
            orgProject.name = project.name
            orgProject.uploaded = project.uploaded
            orgProject.entranceUri = project.entranceUri
            orgProject.type = project.type
            orgProject.linkUrl = project.linkUrl
        } else {
            projects?.add(0, project)
        }
        writeData(data)
        return 1
    }

    override fun findAll(): List<Project> {
        val data = readData()
        if(data.projects != null) {
            return data.projects!!
        }
        return listOf()
    }

    override fun delete(id: String?): Int {
        val data = readData()
        val projects = data.projects
        var count = 0
        if(projects != null) {
            val it = projects.iterator()
            while (it.hasNext()) {
                val one = it.next()
                val pid = one.id
                if (StringUtils.equals(id, pid)) {
                    it.remove()
                    count++
                }
            }
        }
        return count
    }

    private fun doGetProjectById(data: ProjectDataWrapper, id: String?): Project? {
        val projects = data.projects
        if(projects != null) {
            for (one in projects) {
                if (StringUtils.equals(id, one.id)) {
                    return one
                }
            }
        }
        return null
    }

    private fun writeData(data: ProjectDataWrapper) {
        val file = projectMetaJsonFile
        doWriteJsonFile(file, data)
    }

    private fun readData(): ProjectDataWrapper {
        val file = projectMetaJsonFile
        try {
            FileReader(file).use { reader ->
                val jsonReader = JSONReader(reader)
                val data = jsonReader.readObject(ProjectDataWrapper::class.java)
                if (data.projects == null) {
                    data.projects = ArrayList()
                }
                return data
            }
        } catch (e: FileNotFoundException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private val projectMetaJsonFile: File
        private get() {
            val workDir = configuration.workDir
            val path = workDir + File.separator + PaperProperties.PROJECT_META_FILE_NAME
            val file = File(path)
            if (!file.exists()) {
                val parentFile = file.parentFile
                if (!parentFile.exists()) {
                    parentFile.mkdirs()
                }
                val data = ProjectDataWrapper()
                data.projects = ArrayList()
                doWriteJsonFile(file, data)
            }
            return file
        }

    private fun doWriteJsonFile(file: File, data: ProjectDataWrapper) {
        try {
            FileWriter(file).use { writer ->
                JSONObject.writeJSONString(writer, data,
                        SerializerFeature.WriteNullListAsEmpty,
                        SerializerFeature.PrettyFormat,
                        SerializerFeature.SortField)
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }
}